File: backend/src/main/kotlin/com/example/backend/services/EmployeService.kt
package com.example.backend.services

import com.example.backend.controllers.EmployeDto
import com.example.backend.controllers.RegisterRequest
import com.example.backend.dtos.EmployeNewDto
import com.example.backend.models.Employe
import com.example.backend.models.User
import com.example.backend.repositories.EmployeRepository
import com.example.backend.repositories.UserRepository
import org.springframework.http.ResponseEntity
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import java.time.LocalDateTime
import java.util.*

@Service
class EmployeService(
  private val passwordEncoder: PasswordEncoder, // Injected here
  private val employeRepository: EmployeRepository,
  private val userRepository: UserRepository,
) {

  fun createEmployee(registerRequest: RegisterRequest, employe: EmployeDto): Employe {
    if (userRepository.existsByEmail(registerRequest.email)) {
      return null ?: throw IllegalArgumentException("Email already exists")
    }

    // Create a new user
    var user = User().apply {
      nom = registerRequest.firstName
      prenom = registerRequest.lastName
      telephone = registerRequest.phone
      email = registerRequest.email
//      registrationDate = LocalDateTime.now()
//      createdDate = LocalDateTime.now()
//      updatedDate = LocalDateTime.now()
//      username = registerRequest.email
      supprimer = 0
//      password = passwordEncoder.encode(registerRequest.password)
      fonction = registerRequest.role // Initialize and set a default role
    }

    println("Registering user: $user")
    println(passwordEncoder.encode(registerRequest.password))
    // Save the user
    user = userRepository.save(user)
    var employe = Employe().apply {
      password = passwordEncoder.encode(registerRequest.password)
      this.user = user
      identifiant = employe.identifiant
    }
    return employeRepository.save(employe)
  }

  fun getAllEmployees(): List<Employe> {
    return employeRepository.findAllBySupprimerEquals()
  }

  fun getEmployeeById(id: Long): Optional<Employe> {
    return employeRepository.findById(id.toInt())
  }

  fun updateEmployee(id: Long, updatedEmploye: EmployeNewDto): Employe {
    if (id== 0.0.toLong()){
      var user = userRepository.findById(updatedEmploye.user!!.toInt()).get()
      var employe = Employe().apply {
        this.identifiant = updatedEmploye.identifiant
        this.password = updatedEmploye.password
        this.codebarreId = updatedEmploye.codebarreId
        this.type = updatedEmploye.type
        this.user = user
        this.etat = updatedEmploye.etat
        this.faireReductionMax = updatedEmploye.faireReductionMax!!.toInt()
        this.faireReductionMax = updatedEmploye.faireReductionMax.toInt()
        this.supprimer = 0
      }
      return employeRepository.save(employe)
    }
    else {
      val existingEmploye = employeRepository.findById(id.toInt())
        .orElseThrow { IllegalArgumentException("Employee not found with ID: $id") }
      var user = userRepository.findById(updatedEmploye.user!!.toInt()).get()
      existingEmploye.identifiant = updatedEmploye.identifiant
      existingEmploye.password = updatedEmploye.password
      existingEmploye.codebarreId = updatedEmploye.codebarreId
      existingEmploye.type = updatedEmploye.type
      existingEmploye.user = user
      existingEmploye.etat = updatedEmploye.etat
      existingEmploye.faireReductionMax = updatedEmploye.faireReductionMax!!.toInt()
      existingEmploye.faireReductionMax = updatedEmploye.faireReductionMax.toInt()
      existingEmploye.supprimer = 0
      return employeRepository.save(existingEmploye)
    }

  }

  fun deleteEmployee(id: Long) {
    if (!employeRepository.existsById(id.toInt())) {
      throw IllegalArgumentException("Employee not found with ID: $id")
    }
//    employeRepository.deleteById(id.toInt())
    var employe = employeRepository.findById(id.toInt()).get()
    employe.supprimer=1
    employeRepository.save(employe)
  }
}

File: backend_py/pharmaPython/app/services/employe_service.py
from typing import List, Optional
from sqlalchemy.orm import Session
from fastapi import HTTPException
from passlib.context import CryptContext

from app.models.employe import Employe
from app.models.user import User
from app.repositories.employe_repository import EmployeRepository
from app.repositories.user_repository import UserRepository
from app.schemas.auth_dto import RegisterRequest
from app.schemas.employe_dto import EmployeNewDto

pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")


class EmployeService:
  def __init__(self, db: Session):
    self.db = db
    self.employe_repo = EmployeRepository(db)
    self.user_repo = UserRepository(db)

  def create_employee(self, register_request: RegisterRequest, employe_dto: EmployeDto) -> Employe:
    if self.user_repo.exists_by_email(register_request.email):
      raise HTTPException(status_code=400, detail="Email already exists")

    user = User(
      nom=register_request.first_name,
      prenom=register_request.last_name,
      telephone=register_request.phone,
      email=register_request.email,
      supprimer=0,
      fonction=register_request.role,
    )
    user = self.user_repo.save(user)

    employe = Employe(
      password=pwd_context.hash(register_request.password),
      user_id=user.id,
      identifiant=employe_dto.identifiant,
    )
    return self.employe_repo.save(employe)

  def get_all_employees(self) -> List[Employe]:
    return self.employe_repo.find_all_by_supprimer_equals()

  def get_employee_by_id(self, id: int) -> Optional[Employe]:
    return self.employe_repo.find_by_id(id)

  def update_employee(self, id: int, updated_employe: EmployeNewDto) -> Employe:
    user = self.user_repo.find_by_id(updated_employe.user)
    if not user:
      raise HTTPException(status_code=404, detail="User not found")

    if id == 0:
      employe = Employe(
        identifiant=updated_employe.identifiant,
        password=updated_employe.password,
        codebarre_id=updated_employe.codebarre_id,
        type=updated_employe.type,
        user_id=user.id,
        etat=updated_employe.etat,
        faire_reduction_max=updated_employe.faire_reduction_max,
        supprimer=0
      )
    else:
      employe = self.employe_repo.find_by_id(id)
      if not employe:
        raise HTTPException(status_code=404, detail="Employee not found")

      employe.identifiant = updated_employe.identifiant
      employe.password = updated_employe.password
      employe.codebarre_id = updated_employe.codebarre_id
      employe.type = updated_employe.type
      employe.user_id = user.id
      employe.etat = updated_employe.etat
      employe.faire_reduction_max = updated_employe.faire_reduction_max
      employe.supprimer = 0

    return self.employe_repo.save(employe)

  def delete_employee(self, id: int):
    employe = self.employe_repo.find_by_id(id)
    if not employe:
      raise HTTPException(status_code=404, detail=f"Employee not found with ID: {id}")

    employe.supprimer = 1
    self.employe_repo.save(employe)
